### ISSUE TITLE
on_device_transform

### ISSUE THREAD

**Issue by Thomas-ia** [drprojects/superpoint_transformer#186]
### ✅ Code of conduct checklist

- [x] 🌱 I am using the **_latest version_** of the [code](https://github.com/drprojects/superpoint_transformer/tree/master).
- [x] 👩‍💻 I made sure the bug concerns the official [project's codebase](https://github.com/drprojects/superpoint_transformer/tree/master) and not something I coded myself on top of it. (We only provide support for code we wrote and released ourselves).
- [x] 🔎 I took appropriate **_time_** to investigate the problem before filing an issue, but am unable to solve it myself.
- [x] 📙 I **_thoroughly_** went through the [README](https://github.com/drprojects/superpoint_transformer/blob/master/README.md), but could not find the solution there.
- [x] 📘 I **_thoroughly_** went through the tutorial [slides](media/superpoint_transformer_tutorial.pdf), [notebook](notebooks/superpoint_transformer_tutorial.ipynb), and [video](https://www.youtube.com/watch?v=2qKhpQs9gJw), but could not find the solution there.
- [x] 📗 I **_thoroughly_** went through the [documentation](https://github.com/drprojects/superpoint_transformer/tree/master/docs), but could not find the solution there.
- [x] 📜 I went through the **_docstrings_** and **_comments_** in the [source code](https://github.com/drprojects/superpoint_transformer/tree/master) parts relevant to my problem, but could not find the solution there.
- [x] 👩‍🔧 I searched for [**_similar issues_**](https://github.com/drprojects/superpoint_transformer/issues), but could not find the solution there.
- [x] 🔎 I made sure my bug is related to the [project's codebase](https://github.com/drprojects/superpoint_transformer/tree/master) and is not in fact a sub-dependency issue (e.g.: FRNN installation issue, PyTorch installation issue, PyTorch Geometric not installed with GPU support, ...).
- [x] ⭐ Since I am showing interest in the project, I took the time to give the [repo](https://github.com/drprojects/superpoint_transformer/tree/master) a ⭐ to show support. **Please do, it means a lot to us !**

### 🐛 Describe the bug

I found an interesting thing. When the input data passes the ```on_device_transform``` operation, the result ```pos``` and ```x```  will have a certain degree of randomness. The result will be different every time. Is this a bug?

### 📜 Log output

```shell

```

### 🤖 Steps to reproduce the bug

nag = dataset.on_device_transform(nag.to(device_widget.value))

### 🖥️ Environment

```shell
None
```

**Comment by drprojects** [comment]
Hi @Thomas-ia thanks for your question. Is your `dataset` a training dataset or validation/test dataset ? The former should introduce stochasticity, since we have all sorts of augmentations. The latter should not. Can you please share what your `dataset.on_device_transform` are ?

**Comment by Thomas-ia** [comment]
> Hi [@Thomas-ia](https://github.com/Thomas-ia) thanks for your question. Is your `dataset` a training dataset or validation/test dataset ? The former should introduce stochasticity, since we have all sorts of augmentations. The latter should not. Can you please share what your `dataset.on_device_transform` are ?

I can confirm that the validation set is used. This should be easier to reproduce. Randomness will occur with any data. You can use any data to test it.

**Comment by drprojects** [comment]
Using `notebooks/demo.ipynb` as a starting point for instantiation:

```python
# Recover the dataset split we are interested in
dataset = datamodule.val_dataset

# Load the NAG
nag_1 = dataset[0]
nag_2 = dataset[0]

# Apply on-device transforms
nag_1 = dataset.on_device_transform(nag_1.to(device_widget.value))
nag_2 = dataset.on_device_transform(nag_2.to(device_widget.value))

print(nag_1.equal(nag_2))
# > Returns True
```

```python
# Recover the dataset split we are interested in
dataset = datamodule.train_dataset

# Load the NAG
nag_1 = dataset[0]
nag_2 = dataset[0]

# Apply on-device transforms
nag_1 = dataset.on_device_transform(nag_1.to(device_widget.value))
nag_2 = dataset.on_device_transform(nag_2.to(device_widget.value))

print(nag_1.equal(nag_2))
# > Returns False
```

Like I said above, this is the expected behavior. I cannot reproduce the behavior you suggest with the `on_device_transform` of the validation set. To me, things seem to be working as they should. If not, please provide a proper minimum reproducible example showcasing the issue for us to investigate. Closing this issue in the meantime.
